vm/src/obj/objbytes.rs

```diff
@@ -155,8 +164,110 @@ impl PyBytesRef {
             iterated_obj: obj.into_object(),
         }
     }
+
+    fn add(self, other: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        if let Ok(other) = other.downcast::<PyBytes>() {
+            let rhs = &other.value;
+            let elements: Vec<u8> = self.value.iter().chain(rhs.iter()).cloned().collect();
+            vm.ctx.new_bytes(elements)
+        } else {
+            vm.ctx.not_implemented()
+        }
+    }
+
+    fn contains(self, needle: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        // no new style since objint is not.
+        if objtype::isinstance(&needle, &vm.ctx.bytes_type()) {
+            let result = vec_contains(&self.value, &get_value(&needle));
+            vm.ctx.new_bool(result)
+        } else if objtype::isinstance(&needle, &vm.ctx.int_type()) {
+            let result = self
+                .value
+                .contains(&objint::get_value(&needle).to_u8().unwrap());
+            vm.ctx.new_bool(result)
+        } else {
+            vm.new_type_error(format!("Cannot add {:?} and {:?}", self, needle))
+        }
+    }
+
+    fn getitem(self, needle: PyObjectRef, vm: &VirtualMachine) -> PyObjectRef {
+        if objtype::isinstance(&needle, &vm.ctx.int_type()) {
+            match objint::get_value(&needle).to_i32() {
+                Some(pos) => {
+                    if let Some(idx) = self.value.get_pos(pos) {
+                        vm.ctx.new_int(self.value[idx])
+                    } else {
+                        vm.new_index_error("index out of range".to_string())
+                    }
+                }
+                None => {
+                    vm.new_index_error("cannot fit 'int' into an index-sized integer".to_string())
+                }
+            }
+        } else if objtype::isinstance(&needle, &vm.ctx.slice_type()) {
+            vm.ctx
+                .new_bytes((self.value.get_slice_items(&vm, &needle)).unwrap())
+        } else {
+            vm.new_type_error(format!(
+                "byte indices must be integers or slices, not {}",
+                needle
+            ))
+        }
+    }
+
+    fn capitalize(self, vm: &VirtualMachine) -> PyObjectRef {
+        let mut new = vec![self.value[0].to_ascii_uppercase()];
+        new.extend(&self.value[1..]);
+        vm.ctx.new_bytes(new)
+    }
 }
 
 pub fn get_value<'a>(obj: &'a PyObjectRef) -> impl Deref<Target = Vec<u8>> + 'a {
     &obj.payload::<PyBytes>().unwrap().value
 }
+
+/// return true if b is a subset of a.
+fn vec_contains(a: &Vec<u8>, b: &Vec<u8>) -> bool {
+    let a_len = a.len();
+    let b_len = b.len();
+    for (n, i) in a.iter().enumerate() {
+        if n + b_len <= a_len && *i == b[0] {
+            if &a[n..n + b_len] == b.as_slice() {
+                return true;
+            }
+        }
+    }
+    false
+}
+
+fn format_bytes(chars: &Vec<u8>) -> String {
+    let mut res = String::new();
+    for i in chars {
+        res.push_str(BYTES_REPR[*i as usize])
```

Comment:
I think it'd be better to use a range for this - if between 32 (`' '`) and 126 (`'~'`) then just do `res.push(*i as char)`, if it's `\t`, `\n`, or `\r`, do those representations, otherwise use `format!("\\x{:x}", i)`. Also, the `chars` parameter can be an `&[u8]` instead of an `&Vec<u8>`.

Reply:
yes, it's a bit rough, but I did it for perfomance : 
try it : https://pastebin.com/dgBGgXFr but maybe you think to another implementation

Reply:
Using this, I get `format_bytes2`  faster than `format_bytes`:

```rust
#[macro_use]
extern crate timeit;

fn format_bytes(chars: &[u8]) -> String {
    let mut res = String::with_capacity(chars.len());
    for i in chars {
        res.push_str(BYTES_REPR[*i as usize])
    }
    format!("b'{}'", res)
}

fn format_bytes2(chars: &[u8]) -> String {
    let mut res = String::with_capacity(chars.len());
    // 9 10 13  t n r
    for i in chars {
        match i {
            9 => res.push_str("\\t"),
            10 => res.push_str("\\n"),
            13 => res.push_str("\\r"),
            32..=126 => res.push(*i as char),
            _ => res.push_str(&format!("\\x{:x}", i)),
        }
    }
    format!("b'{}'", res)
}
fn main() {
    let c = vec![98, 10, 120, 13, 9, 97, 13, 111];

    let fb = timeit_loops!(10000, {
        format_bytes(&c);
    });
    println!("format_bytes:  {}", fb);

    let fb2 = timeit_loops!(10000, {
        format_bytes2(&c);
    });
    println!("format_bytes2: {}", fb2);
}

const BYTES_REPR: &'static [&'static str] = &[
    "\\x00", "\\x01", "\\x02", "\\x03", "\\x04", "\\x05", "\\x06", "\\x07", "\\x08", "\\t", "\\n",
    "\\x0b", "\\x0c", "\\r", "\\x0e", "\\x0f", "\\x10", "\\x11", "\\x12", "\\x13", "\\x14",
    "\\x15", "\\x16", "\\x17", "\\x18", "\\x19", "\\x1a", "\\x1b", "\\x1c", "\\x1d", "\\x1e",
    "\\x1f", " ", "!", "\"", "#", "$", "%", "&", "'", "(", ")", "*", "+", ",", "-", ".", "/", "0",
    "1", "2", "3", "4", "5", "6", "7", "8", "9", ":", ";", "<", "=", ">", "?", "@", "A", "B", "C",
    "D", "E", "F", "G", "H", "I", "J", "K", "L", "M", "N", "O", "P", "Q", "R", "S", "T", "U", "V",
    "W", "X", "Y", "Z", "[", "\\", "]", "^", "_", "`", "a", "b", "c", "d", "e", "f", "g", "h", "i",
    "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "{", "|",
    "}", "~", "\\x7f", "\\x80", "\\x81", "\\x82", "\\x83", "\\x84", "\\x85", "\\x86", "\\x87",
    "\\x88", "\\x89", "\\x8a", "\\x8b", "\\x8c", "\\x8d", "\\x8e", "\\x8f", "\\x90", "\\x91",
    "\\x92", "\\x93", "\\x94", "\\x95", "\\x96", "\\x97", "\\x98", "\\x99", "\\x9a", "\\x9b",
    "\\x9c", "\\x9d", "\\x9e", "\\x9f", "\\xa0", "\\xa1", "\\xa2", "\\xa3", "\\xa4", "\\xa5",
    "\\xa6", "\\xa7", "\\xa8", "\\xa9", "\\xaa", "\\xab", "\\xac", "\\xad", "\\xae", "\\xaf",
    "\\xb0", "\\xb1", "\\xb2", "\\xb3", "\\xb4", "\\xb5", "\\xb6", "\\xb7", "\\xb8", "\\xb9",
    "\\xba", "\\xbb", "\\xbc", "\\xbd", "\\xbe", "\\xbf", "\\xc0", "\\xc1", "\\xc2", "\\xc3",
    "\\xc4", "\\xc5", "\\xc6", "\\xc7", "\\xc8", "\\xc9", "\\xca", "\\xcb", "\\xcc", "\\xcd",
    "\\xce", "\\xcf", "\\xd0", "\\xd1", "\\xd2", "\\xd3", "\\xd4", "\\xd5", "\\xd6", "\\xd7",
    "\\xd8", "\\xd9", "\\xda", "\\xdb", "\\xdc", "\\xdd", "\\xde", "\\xdf", "\\xe0", "\\xe1",
    "\\xe2", "\\xe3", "\\xe4", "\\xe5", "\\xe6", "\\xe7", "\\xe8", "\\xe9", "\\xea", "\\xeb",
    "\\xec", "\\xed", "\\xee", "\\xef", "\\xf0", "\\xf1", "\\xf2", "\\xf3", "\\xf4", "\\xf5",
    "\\xf6", "\\xf7", "\\xf8", "\\xf9", "\\xfa", "\\xfb", "\\xfc", "\\xfd", "\\xfe", "\\xff",
];

```

Reply:
of course a match... :+1: